### PR TITLE
Profile configurations for pointwise scheduler in python frontend

### PR DESCRIPTION
**Pull request by rdspring1**
# Summary
This PR explores auto-tuning a Pointwise fusion in the python-frontend.

# Details
- Exposes `PointwiseParams` in python frontend.
- Enables scheduling a fusion with the `Pointwise` scheduler and specific parameter configuration.
- Create `autotune_pointwise.py` to test several parameter configurations then apply `DecisionTreeRegressor`
- Using trained `DecisionTreeRegressor`, iterate over parameter configurations for given input arguments and predict the performance metric. The configuration with the best predicted performance is used to schedule a fusion.

# Orthogonal Changes
- Removes child class requirement for defining `schedule` function in `FusionDefinition`
- Recompiles user schedule if `profile == True` in `FusionDefinition.execute`

# API
1. schedule(SchedulerType) -> schedule fusion with specified heuristic. It creates `HeuristicParams` and `SchedulerEntry`
2. `compute_pointwise_heuristics()` -> It creates `HeuristicParams` and `SchedulerEntry`. A `PointwiseParams` reference is returned so any modifications in python will update the reference.
3. `schedule()` - It schedules the fusion given the `SchedulerEntry` and modified `HeuristicParams`.

# Overview of `autotune_pointwise.py`
1. Define a nvfuser fusion and its pytorch eager mode reference.
4. Profile the CUDA kernel performance by iterating over a set of input arguments and scheduler configurations.
5. Train a regression model to predict the desired performance metric given some input arguments and a scheduler configuration.
6. Measure the performance of the regression model.
- Calculate RMSE of predicted and actual performance on test set.
- Find the configuration with the best performance using regression model.    Then, compare against the heuristic configuration selected by nvfuser.
-  For a specific batch size, gather performance across a range of hidden sizes. Calculate performance for best predicted and nvfuser configurations. Plot a chart comparing performance using `matplotlib`.

The selected performance metric is `effective_bandwidth_gbs`. The empirical scheduler selects the configuration that has the highest predicted `effective_bandwidth_gbs`.

# Machine Learning Decisions
### Why Decision Tree?
1. Decision trees work well with discrete parameters like tensor input shape, vectorization factor, and unroll factor.
2. Decision tress are human-understandable because it is a series of if-then-else statements.
3. Relatively fast - quadradic time `O(num_samples * feature_size)` construction and linear time `O(feature_size)` inference

### Random forest algorithm option
- For better predictive performance, we can easily switch to [RandomForestEnsemble](https://scikit-learn.org/1.5/modules/generated/sklearn.ensemble.RandomForestRegressor.html). It is a series of decision trees, where the algorithm picks the best decision from all the trees.

#### How to create graphic of decision tree?
```python
from sklearn import tree
clf = tree.DecisionTreeRegressor()
clf = clf.fit(inputs, outputs)
tree.plot_tree(clf)
```

### Why performance metric is `effective_bandwidth_gbs`?
- Gpu bandwidth is between 0 and 3350 GB/s on H100, so it is a constrained integer metric.

### Why plot metric is `kernel_runtime_ms`?
- The plot looks less noisy because the y-axis scale is constrained between (0, 2.5) ms. Minor perturbations in running-time is smoothed away.

### CODE REVIEW

**Comment by rdspring1** [comment]
I profiled the pointwise scheduler on a variety of tensor shapes and configurations. Then, trained a `DecisionTreeRegressor` on the profiled data. 10% of the data was saved as the test set and the decision tree performed reasonably well.

### Configurations

```
dtype: float32
dimension sizes: [32, 64, 128, 256, 512, 1024, 2048, 4096, 8192, 16384, 32768]
unroll_factor: [1, 4]
vectorize_factor: [1, 2, 4]
```

### Pointwise fusion definition
```python
    def fusion_func(fd: FusionDefinition):
        t0 = fd.from_pytorch(inputs[0])
        t1 = fd.from_pytorch(inputs[1])
        c0 = fd.define_scalar(3.0)
        t2 = fd.ops.add(t0, t1)
        t3 = fd.ops.mul(t2, c0)
        fd.add_output(t3)
```

### Test Results
```
Test prediction error 10.949866516332932

Test performance [  0.23377048   0.45790396   0.97552128   0.99606554   1.99213108
   3.52242277   6.78800221   6.91607772  13.56815247  25.27945649
  45.11410697  94.59409526   0.57273769   0.93508194   1.87135734
   7.32646334   7.18289517   7.63650248  22.55705348  44.43055989
 167.56668303   1.87016387   1.7622698    1.94974531   6.91607772
   7.48065549  12.86147786  14.09815843  69.84543543  47.29704763
  93.8373425    7.18729645   7.63650248  24.85099113  26.65833594
  47.68157647  79.25451224 118.48149305 111.18168542  87.37182728
  70.44845533  48.09704895  58.94305433  66.45704143  12.86147786
  24.85099113  26.65833594  48.07240907  58.64833906 115.56322968
  71.14279189  13.09114711  25.72295573  26.18229422  62.39185006
  58.35655628  89.53944895 166.37826684 172.49511488  87.78048877
  34.11026627  48.07240907  89.19899477  89.6263065   89.7965
  91.40371849  58.35655628 107.64249211 112.245625    92.56457953
  93.38210474  91.81736056  92.5531673   93.37348418  92.79054911
  65.16482118  80.34019049  89.36889762  90.83963456  74.56284664
  91.44825679  90.09294255  86.88642824 118.48149305  90.22821394
  93.88428464  89.15794355  91.40371849  75.67342871  75.74602563
  93.20099993  93.25022578  92.89393384]
  
Test prediction [  0.22028373   0.47728141   0.93508194   0.97487266   1.90912562
   3.45803886   7.04907921   7.04907921  13.82400449  27.15200882
  41.30164723  89.53944895   0.54546446   0.97487266   1.87016387
   6.91607772   6.91607772   6.91607772  22.55705348  41.30164723
 160.68038099   1.83161584   1.79682411   1.90912562   7.05331799
   7.04907921  13.32916797  13.83215544  61.73509375  47.68157647
  88.86111979   7.04907921   7.18729645  24.43680794  24.43680794
  58.35655628 124.78370013  77.16886719  81.73984538  85.46206056
  72.40535687  45.81901489  47.68157647  65.89701018  13.58386545
  24.85099113  26.65833594  48.07240907  58.94305433 112.245625
  72.23987682  12.21840397  27.15200882  24.85099113  70.68726365
  58.64833906 112.245625   118.48149305 124.82519787  89.03501631
  34.09787155  47.28513101 113.33012379  90.35853876  90.33679181
  92.15585807  58.64833906 118.44410539 111.71112202  92.38267622
  92.77048195  92.01994851  92.34254751  93.42718438  92.85232801
  61.09201985  76.1666741   88.69314036  90.70791928  74.28247971
  93.20818723  89.64102215  91.63802978 127.49638926  85.38429709
  90.88362469  90.88362469  89.64637449  75.42815775  75.58384414
  93.17357179  93.42718438  92.78777145]
```

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
Using the trained decision tree, I create an empirical pointwise scheduler. For a single batch size, I iterate over a range of hidden sizes. I run the pointwise fusion for the input arguments using the empirical and nvfuser pointwise schedulers. Using an `unroll_factor` up to 10 and a `vectorization_factor` up to 4, the empirical scheduler is slightly faster than nvfuser.

pointwise_compare_empirical_batchsize512

**Comment by rdspring1** [comment]
!build

**Comment by naoyam** [comment]
Looks like the autotuning part is built on top of some necessary machinery for tuning heuristics parameters from Python. Does it make sense to have that part as a separate PR?

**Comment by rdspring1** [comment]
> Does it make sense to have that part as a separate PR?

Yes, if we're fine with the overall design. The interface for exposing heuristic parameters was co-designed with the autotuning script. #3118 has an orthogonal design because the parameters for the persistent schedulers are generated differently.

**Comment by naoyam** [comment]
> > Does it make sense to have that part as a separate PR?
> 
> Yes, if we're fine with the overall design. The interface for exposing heuristic parameters was co-designed with the autotuning script. #3118 has an orthogonal design because the parameters for the persistent schedulers are generated differently.

The overall design looks good as a starting point. Or, it's still too early to make any conclusive decision, so I expect we would need to do some iterations of explorations. The Python exposure might also change accordingly, so I think as long as there's no red flag, I think we should just merge this PR so that it would make it easier to do further explorations.

**Review by liqiangxl — APPROVED** [review]
LGTM. Prefer to merge and play with it, if there are issues we change later.
